deepliif/models/base_model.py

```diff
@@ -128,23 +136,51 @@ def get_current_losses(self):
                 errors_ret[name] = float(getattr(self, 'loss_' + name))  # float(...) works for both scalar tensor and float number
         return errors_ret
 
-    def save_networks(self, epoch):
+    def save_networks(self, epoch, save_from_one_process=False):
         """Save all the networks to the disk.
-
+        
         Parameters:
             epoch (int) -- current epoch; used in the file name '%s_net_%s.pth' % (epoch, name)
+            save_from_one_process -- when having multiple processes, whether to save from all processes
+                (effectively overwrite each other) or only 1 process; WARNING saving from 1 process at
+                the moment takes LONGER time!!!
         """
+        def custom_save(save_path, remote_transfer_cmd_module, remote_transfer_cmd_function):
+            if remote_transfer_cmd_module and remote_transfer_cmd_function:
+                exec(f'from {remote_transfer_cmd_module} import {remote_transfer_cmd_function}')
+                exec(f'{remote_transfer_cmd_function}("{save_path}")')
+            else:
+                pass
+
         for name in self.model_names:
             if isinstance(name, str):
                 save_filename = '%s_net_%s.pth' % (epoch, name)
                 save_path = os.path.join(self.save_dir, save_filename)
                 net = getattr(self, 'net' + name)
-
+                
                 if len(self.gpu_ids) > 0 and torch.cuda.is_available():
-                    torch.save(net.module.cpu().state_dict(), save_path)
+                    if save_from_one_process:
+                        local_rank = os.getenv('LOCAL_RANK')
+                        rank = os.getenv('RANK')
+
+                        if local_rank is None and rank is None: # DP
+                            torch.save(net.module.cpu().state_dict(), save_path)
+                            custom_save(save_path, self.remote_transfer_cmd_module, self.remote_transfer_cmd_function)
+                        elif local_rank == '0' or rank == '0': # DDP, rank 0
+                            torch.save(net.module.state_dict(), save_path) # saving net.module.cpu().state_dict() causes the next iter hangs at back propagation
+                            custom_save(save_path, self.remote_transfer_cmd_module, self.remote_transfer_cmd_function)
+                        else: # DDP, rank != 0
+                            pass
+                    else:
+                        torch.save(net.module.cpu().state_dict(), save_path)
+                        custom_save(save_path, self.remote_transfer_cmd_module, self.remote_transfer_cmd_function)
+
                     net.cuda(self.gpu_ids[0])
                 else:
                     torch.save(net.cpu().state_dict(), save_path)
+                    custom_save(save_path, self.remote_transfer_cmd_module, self.remote_transfer_cmd_function)
+
+
```

Comment:
Did you check the save function? Does it work properly? If I'm not wrong, the other time, you mentioned that you have some issues with the save function.

Reply:
@Parmida93 
See this mentioned in the PR request:
> Implemented but not enabled features:
> ...
> * save models from only 1 process rather than all processes (hangs at the next iteration)
>   * note that saving from all processes vs. 1 doesn't make a difference if we simply want to have the models persisted, because all processes will end up with exactly the same parameters after gradient sync

Nothing goes wrong in terms of saving if you run our PR version of the code in the tests we did. However, we might miss some cases by chance. Please do bring it up if you see something weird in model saving using this code.